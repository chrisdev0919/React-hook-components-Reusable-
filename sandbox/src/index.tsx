import React, { useState } from "react";
import ReactDOM from "react-dom";
import { ReusableProvider, createStore } from "../../dist";

const useCounter = createStore(() => {
  return useState(0);
});

const useStep = createStore(() => {
  return useState(0);
});

const useFnObject = createStore(() => {
  const [_, setState] = useState(0);

  const callback = () => {
    console.log('callback')
    setState(10);
  }
  return {
    callback
  };
});

const useMultiply = createStore(() => {
  const [counter] = useCounter();
  const [step] = useStep();

  return counter * step;
})

function Header() {
  const [counter, setCounter] = useCounter();
  const [step, setStep] = useStep();

  return (
    <header>
      <div>Counter: { counter }
        <button onClick={ () => setCounter((prev:number) => prev + 1) }>+</button>
        <button onClick = {() => setCounter((prev:number) => prev - 1)}>-</button>
      </div>
      <div> Step: { step }
        <button onClick={ () => setStep((prev:number) => prev + 1) }> +</button>
        <button onClick = {() => setStep((prev:number) => prev - 1)}> -</button>
      </div>
    </header>
  );
}

function Footer() {
    const counter = useCounter(([val]) => val, (a, b) => Math.abs(a - b) < 2);
  const step = useStep(([val]) => Math.floor(val / 4));

  return (
    <footer>
      <div>Counter evens: { counter }</div>
      <div> Step / 4: { step }</div>
    </footer>
  );
}

const Multiply = () => {
  return <span>{ useMultiply() } </span>;
}

const App = () => {
  const [showMultiply, setShowMultiply] = useState(true);
  const callback = useFnObject(state => state.callback);
  setTimeout(() => callback(), 1000);

  return (
    <div>
      <Header/>
      { showMultiply ? <Multiply/> : null }
      <button onClick = {() => setShowMultiply(prev => !prev)}>toggle</button>
      <Footer/>
    </div>
  );
}

const rootElement = document.getElementById("root");
ReactDOM.render(<ReusableProvider><App /></ReusableProvider >, rootElement);
