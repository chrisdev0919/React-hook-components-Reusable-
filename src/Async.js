import React, { useEffect } from "react";
import { useTodos } from "./todos.state";

export const Async = () => {
  const todos = useTodos();

  useEffect(() => {
    todos.fetch();

    return todos.clear;
  }, []);

  return todos.value === null ? (
    "loading..."
  ) : (
    <ul>
      {todos.value.map(todo => (
        <li key={todo.id}>
          <input type="checkbox" checked={todo.completed} />
          {todo.title}
        </li>
      ))}
    </ul>
  );
};
