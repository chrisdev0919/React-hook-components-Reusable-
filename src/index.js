import React from "react";
import ReactDOM from "react-dom";
import { GlobalStateProvider } from "./useGlobalState";
import One from "./One";
import Two from "./Two";
import Three from "./Three";
import "./styles.css";

const initialState = {
  ui: {
    counter: 1,
    random: 0
  }
};

function App() {
  return (
    <GlobalStateProvider initialState={initialState}>
      <One />
      <Two />
      <Three />
    </GlobalStateProvider>
  );
}

const rootElement = document.getElementById("root");
ReactDOM.render(<App />, rootElement);
