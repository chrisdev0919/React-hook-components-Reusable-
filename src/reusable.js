import { shallowCompare } from './shallow-compare';

export const reuse = (unit) => {
  if (!currentStore) {
    throw new Error('Must provide a store first');
  }

  if (currentUnitKey) {
    currentStore.getUnit(currentUnitKey).addDependency(unit);
  }

  const unitContext = currentStore.getUnit(unit);

  if (!unitContext.cachedValue) {
    // save cursor
    const prevUnitKey = currentUnitKey;
    const prevHookIndex = currentHookIndex;

    // reset cursor
    currentUnitKey = unit;
    currentHookIndex = 0;

    // call
    unitContext.cachedValue = unit();

    // run effects:
    unitContext.effects.forEach(effect => {
      if (effect.cleanup) {
        if (typeof effect.cleanup === 'function') {
          effect.cleanup();
        } else {
          console.error(`Cleanup function ${effect.cleanup} is not a function`)
        }
      };
      effect.cleanup = effect.effectFn();
    })
    unitContext.effects = [];

    // restore cursor
    currentUnitKey = prevUnitKey;
    currentHookIndex = prevHookIndex;

  }

  return unitContext.cachedValue;
}

const defaultReducer = (state, value) => {
  if (typeof value === "function") {
    return value(state);
  } else {
    return value;
  }
};

let currentStore;
let currentUnitKey = null;
let currentHookIndex = 0;

const equality = (prev, next) => prev === next;

export const createStore = () => {
  const store = {
    unitContexts: new Map(),
    getUnit: (unit) => {
      // Lazily define hooks for this unit
      if (!store.unitContexts.has(unit)) {
        const unitContext = {
          unit,
          hooks: [],
          effects: [],
          subscribers: [],
          dependencies: new Map(),
          cachedValue: undefined,
          areEqual: unit.areEqual || equality,
          subscribe: (callback) => {
            unitContext.subscribers.push(callback);
            return () => unitContext.unsubscribe(callback);
          },
          unsubscribe: (callback) => {
            unitContext.subscribers = unitContext.subscribers.filter(sub => sub !== callback);
          },
          addDependency: (unit) => {
            const unitContextDep = store.getUnit(unit);
            if (!unitContext.dependencies.has(unit)) {
              const unsubscribe = unitContextDep.subscribe(unitContext.update);
              unitContext.dependencies.set(unit, unsubscribe);
            }
          },
          update: () => {
            const prevValue = unitContext.cachedValue;
            unitContext.cachedValue = undefined;
            const newValue = reuse(unitContext.unit);

            if (!unitContext.areEqual(prevValue, newValue)) { // TBD - change to shallowCompare
              unitContext.subscribers.forEach(sub => {
                sub(newValue);
              });
            }
          }
        }
        store.unitContexts.set(unit, unitContext);
      }
      return store.unitContexts.get(unit);
    },
    subscribe: (unit, callback) => {
      return store.getUnit(unit).subscribe(callback);
    }
  };
  return store;
};

export const setCurrentStore = store => currentStore = store;

export const reuseState = (initialState) => {
  return reuseReducer(defaultReducer, initialState)
}

export const reuseReducer = (reducer, initialState) => {
  if (!currentUnitKey) {
    throw new Error(`reuseMemo hook cannot be called outside of a reuse statement`);
  }

  if (!currentStore) {
    throw new Error('Must provide a store first');
  }

  const unitContext = currentStore.getUnit(currentUnitKey);

  // If hook doesn't exist for this index, create it
  if (unitContext.hooks.length <= currentHookIndex) {
    const curIndex = currentHookIndex; // For closure
    const setState = (action) => {
      const prevState = unitContext.hooks[curIndex].state;
      const newState = reducer(prevState, action);

      unitContext.hooks[curIndex].state = newState;
      unitContext.update();
    }
    const state = (typeof initialState === 'function') ? initialState() : initialState;

    unitContext.hooks[currentHookIndex] = { state, setState, type: 'state' };
  }
  // Get current hook
  let hook = unitContext.hooks[currentHookIndex];
  currentHookIndex++;

  return [hook.state, hook.setState];
}

export const reuseMemo = (fn, deps) => {
  if (!currentUnitKey) {
    throw new Error(`reuseMemo hook cannot be called outside of a reuse statement`);
  }

  if (!currentStore) {
    throw new Error('Must provide a store first');
  }

  const unitContext = currentStore.getUnit(currentUnitKey);
  // If hook doesn't exist for this index, create it
  if (unitContext.hooks.length <= currentHookIndex) {
    unitContext.hooks[currentHookIndex] = { value: undefined, deps: undefined, type: 'memo' };
  }
  // Get current hook
  let hook = unitContext.hooks[currentHookIndex];
  currentHookIndex++;
  let prevDeps = hook.deps;

  // If deps changed - or no deps
  if (!shallowCompare(prevDeps, deps) || !deps) {
    hook.value = fn();
    hook.deps = deps;
  }
  return hook.value;
}

export const reuseCallback = (fn, deps) => {
  return reuseMemo(() => fn, deps);
}

export const reuseEffect = (effectFn, deps) => {
  if (!currentUnitKey) {
    throw new Error(`reuseMemo hook cannot be called outside of a reuse statement`);
  }

  if (!currentStore) {
    throw new Error('Must provide a store first');
  }

  const unitContext = currentStore.getUnit(currentUnitKey);
  // If hook doesn't exist for this index, create it
  if (unitContext.hooks.length <= currentHookIndex) {
    unitContext.hooks[currentHookIndex] = {
      deps: undefined,
      effectFn,
      cleanup: undefined,
      type: 'effect'
    };
  }
  // Get current hook
  let hook = unitContext.hooks[currentHookIndex];
  currentHookIndex++;
  let prevDeps = hook.deps;

  // If deps changed
  if (!shallowCompare(prevDeps, deps) || !deps) {
    unitContext.effects.push(hook);
    hook.deps = deps;
    hook.effectFn = effectFn;
  }
  return;
}


export const Memo = (unit, areEqual = shallowCompare) => {
  unit.areEqual = areEqual;

  return unit;
};

export const reuseRef = (initialVal) => {
  if (!currentUnitKey) {
    throw new Error(`reuseRef hook cannot be called outside of a reuse statement`);
  }

  if (!currentStore) {
    throw new Error('Must provide a store first');
  }

  const unitContext = currentStore.getUnit(currentUnitKey);
  // If hook doesn't exist for this index, create it
  if (unitContext.hooks.length <= currentHookIndex) {
    console.log('create ref');
    unitContext.hooks[currentHookIndex] = {
      ref: {current: initialVal},
      type: 'ref'
    };
  }
  // Get current hook
  let hook = unitContext.hooks[currentHookIndex];
  currentHookIndex++;

  return hook.ref;
} // reuseRef

