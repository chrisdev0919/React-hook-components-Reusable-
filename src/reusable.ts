export type HookFn<HookValue> = () => HookValue;
export type StoreValueChangeCallback<HookValue> = (value: HookValue | null) => void;
export type StoresChangeCallback = () => void;

export class Store<HookValue> {
  subscribers: StoreValueChangeCallback<HookValue>[] = [];
  cachedValue: HookValue | null = null;
  constructor(private fn: HookFn<HookValue>) {
  }

  getValue() {
    return this.cachedValue;
  }

  run() {
    this.cachedValue = this.fn();
    return this.cachedValue;
  }

  subscribe(callback: StoreValueChangeCallback<HookValue>) {
    this.subscribers = [...this.subscribers, callback];
    return () => {
      this.subscribers = this.subscribers.filter(sub => sub !== callback)
    }
  }

  notify() {
    this.subscribers.forEach(sub => sub(this.cachedValue));
  }
}

export class Container {
  stores = new Map<HookFn<any>, Store<any>>();
  subscribers: StoresChangeCallback[] = [];
  onStoresChanged(callback: StoresChangeCallback) {
    this.subscribers = [...this.subscribers, callback];
    return () => {
      this.subscribers = this.subscribers.filter(item => item !== callback);
    }
  }
  createStore(fn: HookFn<any>) {
    if (this.stores.has(fn)) {
      throw new Error('Store already exist');
    }
    const store = new Store(fn);
    this.stores.set(fn, store);
    this.notifyStoresChanged();
  }
  getStore(fn: HookFn<any>):Store<any> {
    if (!this.stores.has(fn)) {
      throw new Error('Store doesn\'t exist');
    }
    return <Store<any>>this.stores.get(fn);
  }
  notifyStoresChanged() {
    this.subscribers.forEach(sub => sub());
  }
  getStoresArray() {
    const storesArray: Store<any>[] = [];

    this.stores.forEach((store) => {
      storesArray.push(store);
    })
    return storesArray;
  }
}
export const createContainer = () => new Container();
let defaultContainer = new Container();
export const getContainer = () => defaultContainer;
export const replaceContainer = (mockedContainer: Container) => defaultContainer = mockedContainer;
