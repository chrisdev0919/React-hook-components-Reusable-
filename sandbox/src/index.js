import React, { useState } from "react";
import ReactDOM from "react-dom";
import { ReusableProvider, reusable } from "../../dist/reusable";

const useCounter = reusable(() => {
  return useState(0);
});

const useStep = reusable(() => {
  return useState(0);
});

const useMultiply = reusable(() => {
  const [counter] = useCounter();
  const [step] = useStep();

  return counter * step;
})

function Header() {
  const [counter, setCounter] = useCounter();
  const [step, setStep] = useStep();

  return (
    <div>
      <div>Counter: {counter}
        <button onClick={() => setCounter(prev => prev + 1)}>+</button>
        <button onClick={() => setCounter(prev => prev - 1)}>-</button>
      </div>
      <div>Step: {step}
        <button onClick={() => setStep(prev => prev + 1)}>+</button>
        <button onClick={() => setStep(prev => prev - 1)}>-</button>
      </div>
    </div>
  );
}

function Footer() {
  const [counter] = useCounter();
  const [step] = useStep();

  return (
    <div>
      <div>Counter: {counter}
      </div>
      <div>Step: {step}
      </div>
    </div>
  );
}

const Multiply = () => {
  return <span>{useMultiply()}</span>;
}

const App = () => {
  const [showMultiply, setShowMultiply] = useState(true);

  return (
    <div>
      <Header/>
      {showMultiply ? <Multiply/> : null }
      <button onClick={() => setShowMultiply(prev => !prev)}>toggle</button>
      <Footer/>
    </div>
  );
}

const rootElement = document.getElementById("root");
ReactDOM.render(<ReusableProvider><App /></ReusableProvider>, rootElement);
