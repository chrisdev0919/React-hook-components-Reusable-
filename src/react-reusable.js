import React, { useState, useContext, useEffect } from "react";
import {shallowEqual} from './shallow-equal';
import {getStore} from './reusable';

const ReusableContext = React.createContext();
export const ReusableProvider = ({ children }) => {
  return (
    <ReusableContext.Provider value={getStore()}>
      <React.Fragment>
        <Units/>
        {children}
      </React.Fragment>
    </ReusableContext.Provider>
  );
};

const Unit = ({ unit }) => {
  unit.run();

  useEffect(() => unit.notify(), [unit.cachedValue]);

  return null;
}

const useStore = () => useContext(ReusableContext);

const Units = () => {
  const store = useStore();
  const [units, setUnits] = useState(() => store.getUnitsArray());

  useEffect(() => {
    return store.onUnitsChanged(() => {
      setUnits(store.getUnitsArray());
    });
  }, []);

  return (
    <React.Fragment>
      {units.map((unit, index) => <Unit key={index} unit={unit}/>)}
    </React.Fragment>
  )
}
const identity = val => val;
const useReuse = (fn, selector = identity, areEqual = shallowEqual) => {
  const unit = useStore().getUnit(fn);
  const [localCopy, setLocalCopy] = useState(() => selector(unit.getValue()));
  
  useEffect(() => {
    return unit.subscribe((newValue) => {
      const selectedNewValue = selector(newValue);
      if (!areEqual(selectedNewValue, localCopy)) {
        setLocalCopy(selectedNewValue);
      }
    });
  }, [unit, selector, areEqual]);

  return localCopy;
}

export const reusable = (fn) => {
  getStore().createUnit(fn);

  return (selector, areEqual) => useReuse(fn, selector, areEqual);
}

// TBD:
// export const reusableReducer = (reducer, initialValue, init, options) => {
//   const fn = () => useReducer(reducer, initialValue, init);

//   getStore().createUnit(fn, options);
//   return (selector, areEqual) => useReuse(fn, selector, areEqual);
// };

// export const reusableState = (init, options) => {
//   const fn = () => useState(init);

//   getStore().createUnit(fn, options);
//   return (selector, areEqual) => useReuse(fn, selector, areEqual);
// };
