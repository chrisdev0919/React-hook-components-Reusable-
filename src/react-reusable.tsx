import * as React from 'react';
import { FunctionComponent, useState, useContext, useEffect } from 'react';
import { shallowEqual, AreEqual } from './shallow-equal';
import { Store, getStore, Unit as UnitClass, HookFn } from './reusable';

const ReusableContext = React.createContext<Store | null>(null);

export const ReusableProvider: FunctionComponent<{}> = ({ children }) => {
  return (
    <ReusableContext.Provider value={getStore()}>
      <React.Fragment>
        <Units />
        {children}
      </React.Fragment>
    </ReusableContext.Provider>
  );
};

const Unit = ({ unit }: { unit: UnitClass<any> }) => {
  unit.run();

  useEffect(() => unit.notify(), [unit.cachedValue]);

  return null;
}

const useStore = () => {
  const store = useContext(ReusableContext) as Store;

  if (store === null) {
    throw new Error('Are you trying to use Reusable without a ReusableProvider?');
  }

  return store;
}

const Units = () => {
  const store = useStore();
  const [units, setUnits] = useState(() => store.getUnitsArray());

  useEffect(() => {
    return store.onUnitsChanged(() => {
      setUnits(store.getUnitsArray());
    });
  }, []);

  return (
    <React.Fragment>
      {units.map((unit: UnitClass<any>, index: number) => <Unit key={index} unit={unit} />)}
    </React.Fragment>
  )
}

type SelectorFn<HookValue> = (val: HookValue) => any;
const identity = (val: any) => val;
function useReuse<HookValue>(
  fn: HookFn<HookValue>,
  selector = identity,
  areEqual = shallowEqual
) {
  const unit = useStore().getUnit(fn);
  const [localCopy, setLocalCopy] = useState(() => selector(unit.getValue()));

  useEffect(() => {
    return unit.subscribe((newValue: any) => {
      const selectedNewValue = selector(newValue);
      if (!areEqual(selectedNewValue, localCopy)) {
        setLocalCopy(() => selectedNewValue);
      }
    });
  }, [unit, localCopy, selector, areEqual]);

  return localCopy;
}

export function reusable<HookValue>(fn: HookFn<HookValue>) {
  getStore().createUnit(fn);

  return (selector?: SelectorFn<HookValue>, areEqual?: AreEqual<HookValue>) => useReuse(fn, selector, areEqual);
}

// TBD:
// export const reusableReducer = (reducer, initialValue, init, options) => {
//   const fn = () => useReducer(reducer, initialValue, init);

//   getStore().createUnit(fn, options);
//   return (selector, areEqual) => useReuse(fn, selector, areEqual);
// };

// export const reusableState = (init, options) => {
//   const fn = () => useState(init);

//   getStore().createUnit(fn, options);
//   return (selector, areEqual) => useReuse(fn, selector, areEqual);
// };
