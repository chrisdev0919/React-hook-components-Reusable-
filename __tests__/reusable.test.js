import {useState} from 'react'
import {renderHook, act} from 'react-hooks-testing-library'
import {getStore, createStore, replaceStore} from '../src/reusable';
import {reusable, ReusableProvider} from '../src/react-reusable';

let store;
let unit;
const value = {title: 'Yo'};
const expectToBeAStore = (object) => {
  expect(object.units).toBeDefined();
}
const expectToBeAUnit = (object) => {
  expect(object.cachedValue).toBeDefined();
}
describe('public store API', () => {
  beforeEach(() => {
    store = getStore();
  });
  it('should allow to get a store', () => {
    expectToBeAStore(store);
  });
  it('should allow to create a store', () => {
    expectToBeAStore(createStore());
  });
  it('should allow to replace the store', () => {
    const newStore = createStore();
    replaceStore(newStore);
    expect(getStore()).toBe(newStore);
  });
  it('should allow to create a unit', () => {
    const fn = () => {};
    store.createUnit(fn);
    expectToBeAUnit(store.getUnit(fn));
  });
  it('should allow to subscribe to creating a unit', () => {
    const fn = () => {};
    let result;
    store.onUnitsChanged(() => {
      result = store.getUnit(fn);
    });
    store.createUnit(fn);
    expectToBeAUnit(result);
  });
  it('should not allow to get a unit without creating it', () => {
    const fn = () => {};

    expect(() => store.getUnit(fn)).toThrow();
  });
  it('should not allow to create a unit twice', () => {
    const fn = () => {};
    store.createUnit(fn);
    expect(() => store.createUnit(fn));
  });
  it('should allow to unsubscribe from subscribing to creating a unit', () => {
    expect.assertions(0);
    const fn = () => {};
    const unsubscribe = store.onUnitsChanged(() => {
      expect(true).toBe(false); // should not reach here
    });
    unsubscribe();
    store.createUnit(fn);
  });

});
describe('public unit API', () => {
  beforeEach(() => {
    const fn = () => value;
    store.createUnit(fn);
    unit = store.getUnit(fn);
  });

  it('should allow to get the cached value', () => {
    expect(unit.cachedValue).toBeDefined();
  });

  it('should allow to run the unit', () => {
    expect(unit.run()).toBe(value);
  });

  it('should cache the unit value', () => {
    unit.run();
    expect(unit.cachedValue).toBe(value);
  });
  it('should allow to subscribe to value change', () => {
    let result;
    unit.subscribe((cachedValue) => {
      result = cachedValue;
    });
    unit.run();
    unit.notify();
    expect(result).toBe(value);
  });
  it('should allow to unsubscribe to value change', () => {
    expect.assertions(0);
    const unsubscribe = unit.subscribe((cachedValue) => {
      expect(cachedValue).toBe(value);
    });
    unit.run();
    unsubscribe();
    unit.notify();
  });
});

describe('reusable', () => {
  beforeEach(() => {
    store = createStore();
    replaceStore(store);
  })
  it('should allow to define a reusable unit', () => {
    const useSomething = reusable(() => useState(0));

    expect(typeof useSomething).toBe('function');
  });
  it('should return a unit value', () => {
    const useSomething = reusable(() => useState(1));
    const { result } = renderHook(useSomething, {
      wrapper: ReusableProvider
    });

    expect(result.current[0]).toBe(1);
  });
  it('should allow to set values on a unit', () => {
    const useSomething = reusable(() => useState(1));
    const {result} = renderHook(useSomething, {
      wrapper: ReusableProvider
    });

    act(() => result.current[1](3));
    expect(result.current[0]).toBe(3);
  });
  it('should allow to set values twice (stale state bug)', () => {
    const useSomething = reusable(() => useState(false));
    const {result} = renderHook(useSomething, {
      wrapper: ReusableProvider
    });

    act(() => result.current[1](prev => !prev));
    expect(result.current[0]).toBe(true);
    act(() => result.current[1](prev => !prev));
    expect(result.current[0]).toBe(false);
  });
  it('should share state between units', () => {
    const useSomething = reusable(() => useState(1));
    const {result} = renderHook(useSomething, {
      wrapper: ReusableProvider
    });

    const {result: result2} = renderHook(useSomething, {
      wrapper: ReusableProvider
    });

    act(() => result.current[1](3));
    expect(result.current[0]).toBe(3);
    expect(result2.current[0]).toBe(3);
  });
  it('should fail without a provider', () => {
    const useSomething = reusable(() => useState(1));
    const {result} = renderHook(useSomething);

    expect(result.error.toString()).toMatchSnapshot();
  });
});
