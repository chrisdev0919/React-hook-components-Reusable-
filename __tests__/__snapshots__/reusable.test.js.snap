// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reusable should fail without a provider 1`] = `"Error: Are you trying to use Reusable without a ReusableProvider?"`;
